support/src/main/java/org/jenkinsci/plugins/workflow/support/steps/MilestoneStepExecution.java

```diff
@@ -0,0 +1,443 @@
+package org.jenkinsci.plugins.workflow.support.steps;
+
+import static java.util.logging.Level.WARNING;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.Map.Entry;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.annotation.CheckForNull;
+
+import org.jenkinsci.plugins.workflow.actions.LabelAction;
+import org.jenkinsci.plugins.workflow.actions.ThreadNameAction;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+import org.jenkinsci.plugins.workflow.graph.BlockEndNode;
+import org.jenkinsci.plugins.workflow.graph.BlockStartNode;
+import org.jenkinsci.plugins.workflow.graph.FlowGraphWalker;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.graph.FlowStartNode;
+import org.jenkinsci.plugins.workflow.steps.AbstractStepExecutionImpl;
+import org.jenkinsci.plugins.workflow.steps.FlowInterruptedException;
+import org.jenkinsci.plugins.workflow.steps.StepContext;
+import org.jenkinsci.plugins.workflow.steps.StepContextParameter;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.DoNotUse;
+
+import com.google.inject.Inject;
+
+import hudson.Extension;
+import hudson.XmlFile;
+import hudson.model.Executor;
+import hudson.model.InvisibleAction;
+import hudson.model.Job;
+import hudson.model.Result;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.model.listeners.RunListener;
+import jenkins.model.CauseOfInterruption;
+import jenkins.model.Jenkins;
+
+public class MilestoneStepExecution extends AbstractStepExecutionImpl {
+
+    private static final Logger LOGGER = Logger.getLogger(MilestoneStepExecution.class.getName());
+
+    @Inject(optional=true) private transient MilestoneStep step;
+    @StepContextParameter private transient Run<?,?> run;
+    @StepContextParameter private transient FlowNode node;
+    @StepContextParameter private transient TaskListener listener;
+
+    private static Map<String, Map<Integer, Milestone>> milestonesByOrdinalByJob;
+
+    @Override
+    public boolean start() throws Exception {
+        if (step.getLabel() != null) {
+            node.addAction(new LabelAction(step.getLabel()));
+        }
+        Integer ordinal = processOrdinal();
+        tryToPass(run, getContext(), ordinal);
+        return true;
```

Comment:
Could use `AbstractSynchronousStepExecutionImpl`, since we unconditionally return true from `start` (never block).